staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go

```diff
@@ -285,6 +285,16 @@ type ValidationRule struct {
 	// e.g. for attribute `foo.34$` appears in a list `testList`, the fieldPath could be set to `.testList['foo.34$']`
 	// +optional
 	FieldPath string `json:"fieldPath,omitempty" protobuf:"bytes,5,opt,name=fieldPath"`
+
+	// optionalOldSelf is used to indicate a transition rule is always evaluated.
+	// If enabled:
+	// - During a create, `oldSelf` evaluates to `null`
+	// - During an update, `oldSelf` evaluates to `null` when the old object is
```

Comment:
Does it break CEL type-checking if we're declaring `oldSelf` is of the type of the containing property here

https://github.com/kubernetes/kubernetes/blob/6b8b1881491e26af2370f08d951b7a816d077665/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L161-L163

but then `oldSelf` can sometimes be null? Do we have good tests for the type behavior of rules which use `oldSelf` and allow it to be null? We should do this on properties of every type we support (int, float, string, bool, slice, object) and evaluate the expression with oldSelf present and absent.

Will a check like `oldSelf == null ? ... : ...` work for both null and non-null `oldSelf` values?

Is there a canonical way rules should check for existence of `oldSelf`?
* `oldSelf == null`?
* `has(oldSelf)`?
* `type(oldSelf) == null_type`?

Can we document that for them (and test all the ways we think it *should* work)?

Reply:
The only portable way to check this condition that works on all types of `oldSelf` is `type(oldSelf) == null_type`. 

- `has(oldSelf)` is a syntax error since `has` is only implemented for map properties, not top level variables. 
- `oldSelf == null` works as expected for a few types like `object` but is a type-checking error when `oldSelf`/`self` is defined as `string` or `integer`, etc. in the OpenAPI schema.
-  `type(oldSelf) == null_type` is universally well-typed, and works as expected for both nullable and non-nullable types.

I've added to the field doc to suggest users to use the `type` form in their CEL expressions. I've also added a test that it works for all OpenAPI schema types.

Reply:
Ok, so for null checks inside the rule, seems like we have two paths:
* types where self is nullable (object? slice? map?) can do `oldSelf == null` or `type(oldSelf) == null_type`, which is nice
* types where self is not nullable (primitive types?) have to do the `type(oldSelf) == null_type` which is ugly, but at least type checking will bark at them if they do `oldSelf == null` at validation time

I do think it's **very** likely people will set `optionalOldSelf: true` and then not guard `oldSelf` appropriately in their rule, exposing themselves to evaluation errors when `oldSelf` is `null`, since all the validation-time type-checking is done with `oldSelf` being of the same type as `self`. That's the risk I'm trying to think if there's any way we can mitigate. I wish we had a known good value for `self` so we could test if their rule errors when evaluated with `self=$good` and `oldSelf=null`.

Reply:
> That's the risk I'm trying to think if there's any way we can mitigate.

this feels cheap and dirty, but... what if we required the rule to contain one of the following (ignoring whitespace) if they opt into evaluation with null oldSelf, just as a sanity check that they are aware of the possibility of null oldSelf in their rule?
* `oldSelf == null`
* `oldSelf != null`
* `type(oldSelf) == null_type`
* `type(oldSelf) != null_type`

Reply:
once that's updated, be sure to update the API doc as well (for this field and for the `rule` field)

Reply:
Updated. And updated API docs and added note to rule field

Reply:
Just caught up on the latest code. The use of CEL Optionals here is really nice.